File: Exceptions/ReallyBadCode/ThrowIndexOutOfRangeExceptionFromCode.vb
Public Module ThrowIndexOutOfRangeExceptionFromCode
	Public Function AccessArray(ByVal values As Integer(), ByVal index As Integer) As Integer
		If index < values.GetLowerBound(0) OrElse index > values.GetUpperBound(0) Then
			Throw New IndexOutOfRangeException("Bad index value.")
		End If

		Return values(index)
	End Function
End Module

File: Exceptions/CodeToImprove/ThrowIndexOutOfRangeExceptionFromCode.cs
using System;

namespace CodeToImprove
{
	public static class ThrowIndexOutOfRangeExceptionFromCode
	{
		public static int AccessArray(int[] values, int index)
		{
			if(index < values.GetLowerBound(0) ||  index > values.GetUpperBound(0))
			{
				throw new IndexOutOfRangeException("Bad index value.");
			}

			return values[index];
		}
	}
}